Create rule S7119: Global should not depend on possibly not yet initialized variables CPP-5655 (#4387)
SonarSource/rspec: rules/S7119/cfamily/rule.adoc

This rule raises an issue if the initializer of the global variable,
refers to another global variable that may not yet be initialized.

== Why is this an issue?

{cpp} allows global variables, to have arbitrally complex initializers.

Such variables are zero-initialized statically,

> and further initialization is performed at runtime,

Reviewer: I would add this if it would be true. However, deferred dynamic initialization is allowed, and variables may be initialized after the star of `main` and before first use.

in the process referred to as dynamic initialization.

The dynamic initialization of such variables is performed in the order of their definitions in the source file.
However, the order of initialization of variables defined in different source files is not specified.
In case when the global variable refers to a variable defined in a different source file,
the behavior of the program is not guaranteed:

[source,cpp]
----
extern int global1;
// Defined in different source file as:
// int global1 = 20;

int const global2 = global1 + 10; // Noncomplaint
----

Depending on the ordering of the initialization of `global1` and `global2` variables,
the `global2` may be set to:

* `30` if `global1` is initialized before `global2`
* `10` if `global2` is initialized before `global1`, as `global1` will have value `0`.

This rule raises an issue if the initializer of the global variable

> directly refers to the other global variable, that may not yet be initialized.

Reviewer: I revised this to follow the style of the introduction sequence, which I agree is much cleaner.

> In other words, referring to other variables is not required to be initialized before a given variable.

Reviewer: I think I wanted to clariy the previous one, but I will use your phasing with names

// Defined in different source file as:
// std::vector<std::string> const list{"entry1", "entry2"};

std::string const firstElem = list[0]; // Noncompliant: undefined behavior
----

=== Static class data members also use dynamic initialization

Static data members of the class are dynamically initialized if the initializer is not a constant expression.
As in the case of global variables, the rule will raise the issue, if static data member of the class with dynamic initialization:
 * refers to global variable of the static data member of the class that may not yet be initialized
 * is referred to by the initializer of the global variable of the static data member of the class in place, when it may not yet be initialized.

[source,cpp]
----
// Clazz.hpp
extern int global;
class Clazz {
  static int staticMem;
};

// Clazz.cpp
#include "Clazz.hpp"

int Clazz::staticMem = global + 10; // Non-complaint: `global` may not yet be initialized

// Source.cpp
#include "Clazz.hpp"

int global = 10;
int otherGlobal = Clazz::staticMem; // Non-complaint: `Clazz::staticMem` may not yet be initalized
----

=== Initialization of variable template instantiations is unordered

> The {cpp}14 introduced the possibility to define a template for the global variable that can be instantiated by providing a template argument.

Reviewer: Good point; I am going to rewrite the section. This should also simply how I mention that in how to fix it.

> The {cpp}14 introduced the possibility to define a template for the global variable that can be instantiated by providing a template argument.

Reviewer: Rewrite done

----

==== Compliant solution

[source,cpp,diff-id=1,diff-type=compliant]
----
// header.h
extern std::string s1;
extern std::string s2;

// source1.cpp
#include "header.h"
std::string s1 = "Some string";
std::string s2 = s1 + " in other file"; // Complaint: `s1` is initialized before `s2`

// source2.cpp
#include "header.h"
----

=== Force constant initialization of referred variable.

Order of initialization issues are limited to variables that are initialized at runtime (has dynamic initialization),
and does not affect variables that are initialized at compile time, i.e., has constant initialization.

This requires the initializer of the variable to be available at compile time,
i.e. use only literals, other constant variables, and invocation of `constexpr` functions or constructors.

==== Noncompliant code example

[source,cpp,diff-id=2,diff-type=noncompliant]
----
// header.h
extern int const count;
extern std::vector<int> entries;

// source1.cpp
#include "header.h"
int const count = 20;

// source2.cpp
#include "header.h"
std::vector<int> entries(count); // Noncomplaint: `count` may not initialized before `entries`
----

==== Compliant solution

> If the project uses {cpp}11 or later standard, you may defined the variable as `constexpr`, to foce constant initialization.

Reviewer: You cannot move it to the header with `constinit` so it does not help with order of initialization. I will mention that in the inline section.

> If the project uses {cpp}11 or later standard, you may defined the variable as `constexpr`, to foce constant initialization.

Reviewer: Added two sections covering mixing `constant` and static function variables or inline variables.

----

==== Handling class static data members

If the static data member is constant, its value may be defined in class:
[source,cpp]
----
struct Clazz {
  static int const constMem = 10;
  static constexpr int constexprMem = 10;
};
----

However, in the case of `const` members and `constexpr` static data members before {cpp}17,
the definition of the class needs to be provided when address or reference to such variable is taken.

Such definition should not repeat the initializer, and can be placed in the source file:
[source,cpp]
----
int const Clazz::constMem;
// separe `Clazz::constexprMem` definition is only required before C++17
int const Clazz::constexprMem; //  const is implied by `constexpr`
----

Or when using {cpp}17 or later in the header file with `inline` keyword.
[source,cpp]
----
inline int const Clazz::constMem;
// separate `Clazz::constexprMem` definition is not required in C++17 or later
----

==== Using `inline` to avoid multiple definitions.

In {cpp} variables declares as `const`, which also covers `constexpr`, has internal linkage.
This means that they are not visible outside of the source file.

> As a consequence, multiple files can define variables with the same name, and each such file will contain an independent copy of the variable.

Reviewer 2: ```suggestion
As a consequence, multiple files can define constants with the same name, and each such file will contain an independent occurrence of the variable.
```

> This also affects cases when the `const` variable is defined in the header file which is included from multiple files.

Reviewer 2: ```suggestion
This also applies when the `const` variable is defined in a header file which is included from multiple files.
```

> In the following example, each `source1.cpp` and `source2.cpp` files contains independent copies of variables `count`.

Reviewer: We said that variables defined before are initialized before, so for me it implied it. Do you have any specific suggestion on how to capture that the copy in each file is initialized at the point of definition so there is no order of initialization fiasco?

> In the following example, each `source1.cpp` and `source2.cpp` files contains independent copies of variables `count`.

Reviewer: Addressed this and the question regarding how this is relevant together.

[source,cpp]
----
// header.h
constexpr int count = 20;

// source1.cpp
#include "header.h"

void print1() {
  std::cout << &count << std::endl;
}

// source2.cpp
#include "header.h"

void print2() {
  std::cout << &count << std::endl;
}
----

This behavior usually does not affect the code, as the value read from the variable is always the same because it is `const`.

> However, taking the address of such a variable will produce different results (i.e., `print1` and `print2` will produce different outputs).

Reviewer 2: I'm not sure how this is related to the current rule...

Reviewer: This is related to the given solution. You cannot make variable const and non-inline if you rely on the address on it, as it will break.

Reviewer: Reworked the comment to make a connection to existing code more clear.

=== Replace global variable with static function variable

A static variable defined in the function body is initialized when the function is called for the first time,
so it is not possible to read its value before the initialization.
In consequence, replacing a global variable with a function that declares a static variable and returns a reference to it eliminates the order of initialization issues.

==== Noncompliant code example

[source,cpp,diff-id=3,diff-type=noncompliant]
----
// header.h
extern std::string s1;
extern std::string s2;

// source1.cpp
#include "header.h"
std::string s1 = "Some string";

// source2.cpp
#include "header.h"
std::string s2 = s1 + " in other file"; // Noncompliant: `s1` may be initialized after `s2`
----

==== Compliant solution

[source,cpp,diff-id=3,diff-type=compliant]
----
// header.h
std::string& getS1();
extern std::string s2;

// source1.cpp
#include "header.h"
std::string& getS1() {
  static std::string s1 = "Some string";
  return s1;
}

// source2.cpp
#include "header.h"

std::string s2 = getS1() + " in other file"; // Compliant: `s1` is initialized as part of `getS1()` call
----

While the above is sufficient to fix the issue,

> replacing `s2` with the `getS2` function defined in a similar way would prevent future problems.

Reviewer 2: ```suggestion
replacing `s2` with a `getS2` function defined in a similar way would prevent future problems.
```

This solution is also applicable to the variable templates and static data members of class templates.

> [source,cpp,diff-id=4,diff-type=noncompliant]

Reviewer 2: The 2 codes might differ too much for the diff to be effective.

Reviewer: Yes, I will remove the diff from here.

----
template<typename CharT>
std::basic_string<CharT> const basicBuildID = /* runtime initializer */;

std::string const buildID = basicBuildID<char>(); // Noncompliant: `basicBuildID<char>` may not be initialized
----

==== Compliant solution

[source,cpp,diff-id=4,diff-type=compliant]
----
template<typename CharT>
std::basic_string<CharT> const& getBuildID() {
  static std:basic_stirng<CharT> basicBuildID =  /* runtime initializer */;
  return basicBuildID;
}

std::string const buildID = getBuildID<char>(); // Compliant: `basicBuildID` is initialized as part of `getBuildID<char>()` call
----

=== Define variables as inline in the header

> With the introduction of `inline` variables in the {cpp}17, non-local variables can now be defined in the header without causing double-definition errors.

Reviewer 2: ```suggestion
With the introduction of `inline` variables in {cpp}17, non-local variables can now be defined in the header without causing double-definition errors.
```

Dynamic initialization of any such `inline` variables is performed before any variable that is defined consistently after it in all source files.
This is generally true for variables defined in the same headers or in sources that always include a given header.

==== Noncompliant code example

[source,cpp,diff-id=5,diff-type=noncompliant]
----
// header.h
extern std::string s1;
extern std::string s2;

// source1.cpp
#include "header.h"
std::string s1 = "Some string";

// source2.cpp
#include "header.h"
std::string s2 = s1 + " in other file"; // Noncompliant: `s1` may be initialized after `s2`
----

==== Compliant solution

[source,cpp,diff-id=5,diff-type=compliant]
----
// header.h
inline std::string s1 = "Some string";
extern std::string s2;

// source1.cpp
#include "header.h"

// source2.cpp
#include "header.h"
std::string s2 = s1 + " in other file"; // Compliant: `s1` may be initialized after `s2`
----

Again, declaring `s1` in the header file is sufficient for addressing the issue,

> however, changing `s2` to also be declared in the header file will prevent initialization order issues related to them.

Reviewer 2: ```suggestion
however, changing `s2` to also be declared in the header file will prevent initialization order issues related to it.
```

----

If the project is limited to {cpp}98/{cpp}03, constant initialization is only supported for variables of integral types
that are defined as `const` and only use literals or other constants in their initializers.

[source,cpp]
----
// header.h
int const count = 20;
extern std::vector<int> entries;

// source1.cpp
#include "header.h"

// source2.cpp
#include "header.h"
std::vector<int> entries(count); // Compliant: `count` is initialized at compile time to 20
----

==== Handling class static data members

If a static data member is constant, its value may be defined in its class:
[source,cpp]
----
struct Clazz {
  static int const constMem = 10;
  static constexpr int constexprMem = 10;
};
----

However, in the case of `const` members and `constexpr` static data members before {cpp}17,
the definition of the variable needs to be provided when its address is taken or a reference to it is created.

Such a definition should not repeat the initializer, and can be placed in a source file (not in the header):
[source,cpp]
----
int const Clazz::constMem;
// separate `Clazz::constexprMem` definition is only required before C++17

> int const Clazz::constexprMem; //  const is implied by `constexpr`

Reviewer: I was sure tht does not compile, but all compiles seem to accept it now.

replacing `s2` with a `getS2` function defined in a similar way would prevent future problems.

This solution is also applicable to the variable templates and static data members of class templates.

[source,cpp]
----
template<typename CharT>
std::basic_string<CharT> const basicBuildID = /* runtime initializer */;

std::string const buildID = basicBuildID<char>(); // Noncompliant: `basicBuildID<char>` may not be initialized
----

==== Compliant solution

[source,cpp]
----
template<typename CharT>
std::basic_string<CharT> const& getBuildID() {
  static std:basic_stirng<CharT> basicBuildID =  /* runtime initializer */;
  return basicBuildID;
}

std::string const buildID = getBuildID<char>(); // Compliant: `basicBuildID` is initialized as part of `getBuildID<char>()` call
----

==== Constant initialization mitigates initialization overhead

As the static function variable is initialized when the function is first called,
an additional boolean flag needs to be stored to indicate that the variable is already initialized.
Furthermore, if the function is called from multiple threads, a code synchronizing he initialization of such variable is inserted by compiler.

> As a consequence using static variables has a performance impact related to checks if initialization is required,
> however they are not necessary if it is analyzed by compile-time and have constant initialization.

Reviewer 2: ```suggestion
As a consequence, dynamic initialization of static local variables comes with a performance cost that can be avoided if the variable can be initialized constantly, at compile-time.
```

> For constant initialized variables, you can force constant initialization, by using:
> 
> * `constexpr` (since {cpp}11) if the variable is not mutated and has a constant destructor,
> * `continit` (since {cpp}20) otherwise

Reviewer 2: ```suggestion
You can enforce constant initialization, by using:

* `constexpr` (since {cpp}11) if the variable is not mutated and has a constant destructor,
* `continit` (since {cpp}20) otherwise
```

=== Define variables as inline in the header

With the introduction of `inline` variables in {cpp}17, non-local variables can now be defined in the header without causing double-definition errors.
Dynamic initialization of any such `inline` variables is performed before any variable that is defined consistently after it in all source files.
This is generally true for variables defined in the same headers or in sources that always include a given header.

==== Noncompliant code example

[source,cpp,diff-id=5,diff-type=noncompliant]
----
// header.h
extern std::string s1;
extern std::string s2;

// source1.cpp
#include "header.h"
std::string s1 = "Some string";

// source2.cpp
#include "header.h"
std::string s2 = s1 + " in other file"; // Noncompliant: `s1` may be initialized after `s2`
----

==== Compliant solution

[source,cpp,diff-id=5,diff-type=compliant]
----
// header.h
inline std::string s1 = "Some string";
extern std::string s2;

// source1.cpp
#include "header.h"

// source2.cpp
#include "header.h"
std::string s2 = s1 + " in other file"; // Compliant: `s1` may be initialized after `s2`
----

> Again, declaring `s1` in the header file is sufficient for addressing the issue,

Reviewer 2: ```suggestion
Again, declaring `s1` in the header file is sufficient to address the issue,
```

however, changing `s2` to also be declared in the header file will prevent initialization order issues related to it.

==== Use `constinit` for constant initilization

> As mentioned in the previous section, combining `inline` and `constexpr` will prevent the variable from being defined multiple times,
> in different source files. 

Reviewer 2: ```suggestion
As mentioned in the previous section, combining `inline` and `constexpr` will prevent the variable from being defined multiple times in different source files. 
```

> Furthermore, since {cpp}20 you can use `constinit` to enforce constant initialization of global variables,

Reviewer 2: ```suggestion
Furthermore, since {cpp}20 you can use `constinit` to enforce constant initialization of global variables
```

that needs to be mutated.

==== Templates are implicitly `inline`

> Instantiations of variable templates and static data members of class template instantiation, are implicitly `inline`.
> As a consequence, marking them as `inline` does not impact the order of initialization and does not prevent the issue from occurring.

Reviewer 2: ```suggestion
Even though instantiations of variable templates and static data members of class template instantiation are
implicitly `inline`, their order of initialization is not specified, and explicitly marking them inline will have no impact.
```

=== Updating code incrementally

The options for addressing the order of initialization issues for non-`const` global variables are limited,

> especially in cases when older {cpp} standard.

Reviewer 2: ```suggestion
especially when targeting old {cpp} standards.
```

> The most robust solution of refactoring the global variable into a static function variable,
> requires replacing references to variables with the call to newly used functions.

Reviewer 2: ```suggestion
The most robust solution is to replace the global variable with a function returning a reference to a local static variable. This requires replacing all references to the global variable with a function call.
```

----

==== Compliant solution

[source,cpp,diff-id=6,diff-type=compliant]
----
// header.h
std::string& getS1();
static std::string& s1 = getS1();
extern std::string s2;

// source1.cpp
#include "header.h"
std::string& getS1() {
  static std::string var = "Some string";
  return var;
}

// source2.cpp
#include "header.h"
std::string s2 = s1 + " in other file"; // Compliant: `s1` is a reference to `var`, that is initialized as part of `getS1()` call
----

This solution is inspired by "Nifty Counter" idiom. 

== Resources

=== Documentation

* {cpp} reference - https://en.cppreference.com/w/cpp/language/siof[Static Initialization Order Fiasco]
* {cpp} reference - https://en.cppreference.com/w/cpp/language/initialization#Non-local_variables[Initialization of non-local variables]
* {cpp} reference - https://en.cppreference.com/w/cpp/language/zero_initialization[Zero-initialization]
* {cpp} reference - https://en.cppreference.com/w/cpp/language/constant_initialization[Constant initialization]

=== Articles & blog posts

* WikiBooks - https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Nifty_Counter[Nifty Counter]

=== Related rules

> * S5421 detects uses of non-const global variables.

Reviewer 2: ```suggestion
* S5421 detects non-const global variables.
```

// Clazz.cpp
#include "Clazz.hpp"

> int Clazz::staticMem = global + 10; // Noncompliant: "global" may not yet be fully initalized

Reviewer 2: ```suggestion
int Clazz::staticMem = global + 10; // Noncompliant: "global" may not yet be fully initialized
```

// Source.cpp
#include "Clazz.hpp"

int global = 10;

> int otherGlobal = Clazz::staticMem; // Noncompliant: "Clazz::staticMem" may not yet be fully initalized

Reviewer 2: ```suggestion
int otherGlobal = Clazz::staticMem; // Noncompliant: "Clazz::staticMem" may not yet be fully initialized
```

----

=== Initialization of instantiated variable is unordered

If a global variable instantiated from a template requires a dynamic initialization (because its initialization is non-constant),
the ordering of this initialization is unspecified.
It may be performed before the initialization of a global variable declared before it.
As a consequence, such variables should not refer to any dynamically initialized global variable in their initializers, 
or be referred to from the initializer of any global variable with dynamic initialization.

> An instantiated variable may be produced directly from the instantiation of a variable template (introduced in {cpp}14),
> when such a template is used with template arguments:

Reviewer 2: I'm still unsure of the value of the end of this sentence
```suggestion
An instantiated variable may be produced directly from the instantiation of a variable template (introduced in {cpp}14):
```

  return basicBuildID;
}

std::string const buildID = getBuildID<char>(); // Compliant: "basicBuildID" is initialized as part of "getBuildID<char>()" call
----

==== Constant initialization mitigates initialization overhead

> As the static function variable is initialized when the function is first called,

Reviewer 2: ```suggestion
As the static local variable is initialized when the function is first called,
```
